### PR TITLE
Integrate test suite with GitHub Actions

### LINKED ISSUE #1: Test Suite Integration
## Is your feature request related to a problem?

Currently, no tests are run.

## Describe the solution you'd like

Integrate the Bats test suite and all linting targets into GH actions.

## Alternatives you've considered

Let TravisCI run them. Although I'm not a big fan of this.

## What are you going to contribute?

Help with understanding how the test suite should work.

### PR DESCRIPTION
**Pull request by radicand**
This is for integrated CI in GitHub Actions.

Closes #1

### CODE REVIEW

File: test/test_helper/common.bash [hunk 1]
```diff
@@ -4,7 +4,7 @@ load 'test_helper/bats-support/load'
 load 'test_helper/bats-assert/load'
 
 # shellcheck disable=SC2034
-NAME=tvial/docker-mailserver:testing
+# NAME=tvial/docker-mailserver:testing
```
Comment (martin-schulze-vireso) [inline test/test_helper/common.bash:7]: Commenting this out might have bad sideeffects with the change `IMAGE_NAME=${NAME}` in `setup.sh` above.

File: setup.sh [hunk 2]
```diff
@@ -231,18 +231,23 @@ function _main
...
+  # echo "DEBUG: INFO=${INFO}"
+  # echo "DEBUG: PRE_IMAGE_NAME=${IMAGE_NAME}"
+  # echo "DEBUG: PRE_CONTAINER_NAME=${CONTAINER_NAME}"
 
   if [[ -z ${IMAGE_NAME} ]]
   then
     if [[ ${CRI} == "docker" ]]
     then
-      IMAGE_NAME=tvial/docker-mailserver:latest
+      IMAGE_NAME=${NAME}
```
Comment (martin-schulze-vireso) [inline setup.sh:242]: Better to use `${NAME?}` here, to fail early and loud.

**Review by martin-schulze-vireso — CHANGES_REQUESTED** [review]

File: test/test_helper/common.bash (hunk 1 above)
Comment (radicand) [reply test/test_helper/common.bash:7]: it's not obvious, but NAME is seeded in the env in the GHA pipeline (https://github.com/test-org-dm/docker-mailserver/blob/gha/test-suite/.github/workflows/build_test_push.yml#L101)

File: setup.sh (hunk 2 above)
Comment (radicand) [reply setup.sh:242]: Worth a change to look at, again per comment above, NAME should always be present

File: test/test_helper/common.bash (hunk 1 above)
Comment (martin-schulze-vireso) [reply test/test_helper/common.bash:7]: That is okay for CI but doesn't that hamper local development? We now need to set up `$NAME` manually, which is prone to errors like forgetting to do so or typos.

Comment (radicand) [reply test/test_helper/common.bash:7]: This is a valid point in that we should have some sort of fallback default. I'll see what our options are here.

File: setup.sh (hunk 2 above)
Comment (williamdes) [reply setup.sh:242]:
```suggestion
      IMAGE_NAME=${NAME:-tvial/docker-mailserver:latest}
```

Comment (martin-schulze-vireso) [reply setup.sh:242]: @williamdes I am not sure if the default value should be set at this depth. Come to think of it, I am not even sure if `NAME` should be used as an interface here, that will spell trouble somewhere down the line due to name clashes.

File: test/test_helper/common.bash (hunk 1 above)
Comment (radicand) [reply test/test_helper/common.bash]: I've updated this to fallback as discussed.

File: setup.sh (hunk 2 above)
Comment (radicand) [reply setup.sh]: I did end up changing as @williamdes suggested, if it reaches that far into the execution without NAME set, it does either need a default value or to bomb. Again, open to suggestions.

**Review by williamdes — APPROVED** [review]

File: .github/workflows/build_test_push.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+      - name: HADOLINT
+        run: |
+          sudo curl -L https://github.com/hadolint/hadolint/releases/download/v${HADOLINT_VERSION}/hadolint-$(uname -s)-$(uname -m) -o /usr/local/bin/hadolint
+          sudo chmod +rx /usr/local/bin/hadolint
+          make hadolint
+        env:
+          HADOLINT_VERSION: 1.18.0
+      - name: SHELLCHECK
+        run: |
+          sudo wget -qO- "https://github.com/koalaman/shellcheck/releases/download/v${SHELLCHECK_VERSION}/shellcheck-v${SHELLCHECK_VERSION}.linux.x86_64.tar.xz" | tar -xJ
```
Comment (williamdes) [inline .github/workflows/build_test_push.yml:28]: I was about to create a shellcheck github action, do you want me to proceed ?

Comment (radicand) [reply .github/workflows/build_test_push.yml:28]: I don't have strong preferences, I'm just mimicking what the existing build process does; if there's an action driven way that works with the existing lint.sh script, go for it.

**Comment by radicand** [comment]
I've completed the MVP for lint>build>test>publish, marking this ready for review now.

File: .github/workflows/build_test_push.yml [hunk 4]
```diff
@@ -0,0 +1,161 @@
+name: CI-WIP
```
Comment (williamdes) [inline .github/workflows/build_test_push.yml:1]:
```suggestion
name: Build, test and push
```

File: .github/workflows/build_test_push.yml [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+      - name: SHELLCHECK
+        run: |
+          sudo wget -qO- "https://github.com/koalaman/shellcheck/releases/download/v${SHELLCHECK_VERSION}/shellcheck-v${SHELLCHECK_VERSION}.linux.x86_64.tar.xz" | tar -xJ
+          sudo mv "shellcheck-v${SHELLCHECK_VERSION}/shellcheck" /usr/bin/
+          make shellcheck
+        env:
+          SHELLCHECK_VERSION: 0.7.1
+      - name: ECLINT
+        run: |
+          sudo wget -qO- "https://github.com/editorconfig-checker/editorconfig-checker/releases/download/${ECLINT_VERSION}/ec-linux-amd64.tar.gz" | tar -xaz
```
Comment (williamdes) [inline .github/workflows/build_test_push.yml:35]: you could use curl here and above to be consistent with HADOLINT ?

**Review by williamdes — APPROVED** [review]

File: .github/workflows/build_test_push.yml (hunk 3 above)
Comment (williamdes) [reply .github/workflows/build_test_push.yml:28]: Here it is https://github.com/sudo-bot/action-shellcheck#readme
enjoy !
I added make and bash for you ;)

**Comment by aendeavor** [comment]
Nice! There are some comments left that need resolving, but all in all, this looks superb! I will now go ahead back to the main repo and start the migration process so we all can enjoy the new testing system.

@radicand and @williamdes thank you very much for your work here. We will make sure this is honored in the new repository!

**Comment by radicand** [comment]
Just updated to remove WIP.

File: .github/workflows/build_test_push.yml (hunk 5 above)
Comment (radicand) [reply .github/workflows/build_test_push.yml]: Just made this change to be consistently

File: .github/workflows/build_test_push.yml (hunk 3 above)
Comment (radicand) [reply .github/workflows/build_test_push.yml]: Just noting that this is something to re-visit later - as noted, the linter currently uses a custom lint.sh script to manage the linting process, we'd need to ensure the action method works the same way (or accomplished the same outcomes). It's not a blocker to merge, more of an enhancement IMO.

**Review by aendeavor — APPROVED** [review]

**Pull request merged**